build.sc

```diff
@@ -0,0 +1,220 @@
+
+  // Get rid of the `metronome-2.12.10-` part from the artifact name. The JAR name suffix will shows the Scala version.
+  // Check with `mill show metronome[2.12.10].__.artifactName` or `mill __.publishLocal`.
+  private def removeCrossVersion(artifactName: String): String =
+    "metronome-" + artifactName.split("-").drop(2).mkString("-")
+
+  // In objects inheriting this trait, use `override def moduleDeps: Seq[PublishModule]`
+  // to point at other modules that also get published. In other cases such as tests
+  // it can be `override def moduleDeps: Seq[JavaModule]`, i.e. point at any module.
+  trait Publishing extends PublishModule {
+    def description: String
+
+    // Make sure there's no newline in the file.
+    override def publishVersion = versionFile.currentVersion().toString
+
+    override def pomSettings = PomSettings(
+      description = description,
+      organization = "io.iohk",
+      url = "https://github.com/input-output-hk/metronome",
+      licenses = Seq(License.`Apache-2.0`),
+      versionControl = VersionControl.github("input-output-hk", "metronome"),
+      // Add yourself if you make a PR!
+      developers = Seq(
+        Developer("aakoshh", "Akosh Farkash", "https://github.com/aakoshh")
+      )
+    )
+  }
+
+  /** Common properties for all Scala modules. */
+  trait SubModule extends ScalaModule {
+    override def scalaVersion = crossScalaVersion
+    override def artifactName = removeCrossVersion(super.artifactName())
+
+    override def ivyDeps = Agg(
+      ivy"org.typelevel::cats-core:${VersionOf.cats}",
+      ivy"org.typelevel::cats-effect:${VersionOf.cats}"
+    )
+
+    // `extends Tests` uses the context of the module in which it's defined
+    trait TestModule extends Tests {
+      override def artifactName =
+        removeCrossVersion(super.artifactName())
+
+      override def testFrameworks =
+        Seq("org.scalatest.tools.Framework")
+
+      // It may be useful to see logs in tests.
+      override def moduleDeps: Seq[JavaModule] =
+        super.moduleDeps ++ Seq(logging)
+
+      override def ivyDeps = Agg(
+        ivy"org.scalatest::scalatest:${VersionOf.scalatest}",
+        ivy"org.scalacheck::scalacheck:${VersionOf.scalacheck}",
+        ivy"ch.qos.logback:logback-classic:${VersionOf.logback}"
+      )
+
+      def single(args: String*) = T.command {
+        super.runMain("org.scalatest.run", args: _*)
+      }
+    }
+  }
+
+  /** Storage abstractions, e.g. a generic key-value store. */
+  object storage extends SubModule
+
+  /** Emit trace events, abstracting away logs and metrics.
+    *
+    * Based on https://github.com/input-output-hk/iohk-monitoring-framework/tree/master/contra-tracer
+    */
+  object tracing extends SubModule with Publishing {
+    override def description: String =
+      "Abstractions for contravariant tracing."
+  }
+
+  /** Additional crypto utilities such as threshold signature. */
+  object crypto extends SubModule with Publishing {
```

Comment:
I not sure what primitives we will need, but for more advanced crypto stuff we will probably need bouncy castle library

Reply:
I wanted to put it here first to just create some abstractions like `Signature[OverWhatType]` and `GroupSignature[OverWhatType]` or whatever in case we need them until we can get the crypto published from Mantis. 

HotStuff recommends using k-of-n threshold signatures, so unlike the current checkpoint where the group signature is just a list, this would be single signature, but comprise of signatures from the group members. It can work both ways, but it's quicker to check just 1 signature, not 7, and requires less space.

In any case I just wanted to put it down that a module like this, shared by the others, will most likely be necessary.